From fa59cb046fed5736691d4a14a30dfb65ee2e9ba2 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 19/43] UI loop reboot on no GPU platform when go into early
 suspend.

ColorFade app calls opengl-es 2.0 api to draw black.
On android, it will call GLESv2 wrapper which dlsym GPU api symbols.
On no GPU platform, GLESv2 wrapper dlsym itself api symbols.
And the opengl-es 2.0 api will loop call itself on no GPU platform.
To avoid that, opengl-es 1.0 software implementation should check this case.
eglCreateContext should check attrib_list and reject bad match client version.

EGL_NO_CONFIG should define to -1.
Because opengl-es 1.0 software implementation will return 0 as effective config.
---
 opengl/libagl/egl.cpp                              | 32 +++++++++++++++++++++-
 .../surfaceflinger/RenderEngine/RenderEngine.h     |  2 +-
 2 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/opengl/libagl/egl.cpp b/opengl/libagl/egl.cpp
index b79051c49..99c0bcf5a 100644
--- a/opengl/libagl/egl.cpp
+++ b/opengl/libagl/egl.cpp
@@ -1714,11 +1714,41 @@ EGLBoolean eglQuerySurface( EGLDisplay dpy, EGLSurface eglSurface,
 }
 
 EGLContext eglCreateContext(EGLDisplay dpy, EGLConfig config,
-                            EGLContext /*share_list*/, const EGLint* /*attrib_list*/)
+                            EGLContext /*share_list*/, const EGLint* attrib_list)
 {
     if (egl_display_t::is_valid(dpy) == EGL_FALSE)
         return setError(EGL_BAD_DISPLAY, EGL_NO_SURFACE);
 
+    EGLint renderType = 0;
+    if (getConfigAttrib(dpy, config, EGL_RENDERABLE_TYPE, &renderType) == EGL_FALSE) {
+        return setError(EGL_BAD_CONFIG, EGL_NO_SURFACE);
+    }
+
+    EGLint major = 1;
+    if (attrib_list != NULL) {
+        for (EGLint i=0; attrib_list[i]!= EGL_NONE; i+=2) {
+            if (attrib_list[i] == EGL_CONTEXT_CLIENT_VERSION) {
+                major = attrib_list[i+1];
+                break;
+            }
+        }
+    }
+
+    switch (major) {
+        case 1:
+            if (!(renderType & EGL_OPENGL_ES_BIT)) {
+                return setError(EGL_BAD_MATCH, EGL_NO_SURFACE);
+            }
+            break;
+        case 2:
+            if (!(renderType & EGL_OPENGL_ES2_BIT)) {
+                return setError(EGL_BAD_MATCH, EGL_NO_SURFACE);
+            }
+            break;
+        default:
+            return setError(EGL_BAD_MATCH, EGL_NO_SURFACE);
+    }
+
     ogles_context_t* gl = ogles_init(sizeof(egl_context_t));
     if (!gl) return setError(EGL_BAD_ALLOC, EGL_NO_CONTEXT);
 
diff --git a/services/surfaceflinger/RenderEngine/RenderEngine.h b/services/surfaceflinger/RenderEngine/RenderEngine.h
index 954457946..ebcf15ed3 100644
--- a/services/surfaceflinger/RenderEngine/RenderEngine.h
+++ b/services/surfaceflinger/RenderEngine/RenderEngine.h
@@ -26,7 +26,7 @@
 #include <math/mat4.h>
 #include <Transform.h>
 
-#define EGL_NO_CONFIG ((EGLConfig)0)
+#define EGL_NO_CONFIG ((EGLConfig)-1)
 
 // ---------------------------------------------------------------------------
 namespace android {
-- 
2.11.0

